### PR TITLE
RUMM-593 Make the BroadcastReceivers thread safe

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/receiver/ThreadSafeReceiver.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.concurrent.atomic.AtomicBoolean
+
+internal abstract class ThreadSafeReceiver : BroadcastReceiver() {
+
+    val isRegistered = AtomicBoolean(false)
+
+    fun registerForFilter(
+        context: Context,
+        filter: IntentFilter,
+        doOnRegister: ((intent: Intent?) -> Unit)? = null
```
Comment (xgouchet) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/receiver/ThreadSafeReceiver.kt:22]: Why do you want to use a lambda here ? Can't you simply return the intent and let the caller deal with it if needed ? This seams like some overkill design.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/receiver/ThreadSafeReceiver.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import java.util.concurrent.atomic.AtomicBoolean
+
+internal abstract class ThreadSafeReceiver : BroadcastReceiver() {
+
+    val isRegistered = AtomicBoolean(false)
+
+    fun registerForFilter(
```
Comment (xgouchet) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/receiver/ThreadSafeReceiver.kt:19]: This name seems weird. Can you simply call it `registerReceiver` ?

File: dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/receiver/ThreadSafeReceiver.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        val intent = context.registerReceiver(this, filter)
+        doOnRegister?.invoke(intent)
+        isRegistered.set(true)
+    }
+
+    fun checkAndUnregister(context: Context) {
+        if (isRegistered.compareAndSet(true, false)) {
+            context.unregisterReceiver(this)
+        }
+    }
```
Comment (xgouchet) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/receiver/ThreadSafeReceiver.kt:34]: again the `check` part in the name is not necessary, you can simply call it `unregisterReceiver`

File: dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/receiver/ThreadSafeReceiver.kt (hunk 1 above)
Comment (mariusc83) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/receiver/ThreadSafeReceiver.kt:22]: Yeah...it makes sense...I tried several approaches and I guess this is a left over from a different approach.

**Pull request merged**